### DISCUSSION TITLE
Unstable spectrogram loss

### DISCUSSION THREAD

**Discussion by jonashaag** [asteroid-team/asteroid#501] [Q&A] [answered]
When I train a speech enhancement model with spectrogram based loss, I get unstable training and it doesn't seem to converge. When I train the same models with SI-SDR, it works well.

Validation losses (spectrogram loss vs SI-SDR loss):

<img width="293" alt="Screenshot 2021-05-14 at 15 38 13" src="https://user-images.githubusercontent.com/175722/118279085-00389100-b4cb-11eb-91b9-76c45b3edfcd.png">
<img width="306" alt="Screenshot 2021-05-14 at 15 38 37" src="https://user-images.githubusercontent.com/175722/118279095-03338180-b4cb-11eb-88b5-7b64849e0fd6.png">

The loss is `magnitude MSE + lambda x re/im MSE`, along the lines of:

```py
def specloss(est, tar):
    est = stft(est) # Uses torch.stft
    tar = stft(tar)
    magloss = torch.mean((est.abs() - tar.abs()) ** 2)
    reim_loss = torch.mean((est.real - tar.real) ** 2) + torch.mean((est.imag - tar.imag) ** 2)
    lambda_ = 0.5
    return lambda_ * magloss + (1 - lambda_) * reim_loss
```

Things I've tried that don't make a difference:

- Using a different model
- Using a smaller or larger dataset
- Smaller learning rate
- Other batch size
- Other `lambda_` values (even to 0.0 or 1.0)
- Other/mean of multiple STFT sizes and hop sizes
- Using Asteroid STFT as opposed to Torch STFT

Any ideas how I can debug this? Could it be that the gradients are crazy somewhere? (I use gradient clipping.) What's the best way to debug gradients?

Thanks! :-)

---

Re: “doesn’t converge”: this is the PESQ curve of the unstable training. For stable training the PESQ curve is stable and reaches much larger numbers.

![8569AC52-D40B-4D89-B998-C927D848A812](https://user-images.githubusercontent.com/175722/118324153-f3835f80-b501-11eb-8d19-22b32ee177cf.jpeg)

---

What's also interesting is that the gradients and thus the params are distributed a lot differently with the losses.

Spectrogram based loss:

<img width="1019" alt="Screenshot 2021-05-17 at 10 41 06" src="https://user-images.githubusercontent.com/175722/118460135-b33c0100-b6fc-11eb-9500-686437d95fda.png">
<img width="1036" alt="Screenshot 2021-05-17 at 10 41 26" src="https://user-images.githubusercontent.com/175722/118460157-b8994b80-b6fc-11eb-87f7-c03e05dc2e30.png">

SI-SDR:

<img width="1020" alt="Screenshot 2021-05-17 at 10 40 56" src="https://user-images.githubusercontent.com/175722/118460193-bfc05980-b6fc-11eb-85ae-a04d2e309f82.png">
<img width="1039" alt="Screenshot 2021-05-17 at 10 41 36" src="https://user-images.githubusercontent.com/175722/118460206-c2bb4a00-b6fc-11eb-9d68-4cdee5ead9cc.png">

**Comment by jonashaag** [comment]
Re: “doesn’t converge”: this is the PESQ curve of the unstable training. For stable training the PESQ curve is stable and reaches much larger numbers.

![8569AC52-D40B-4D89-B998-C927D848A812](https://user-images.githubusercontent.com/175722/118324153-f3835f80-b501-11eb-8d19-22b32ee177cf.jpeg)

**Comment by mpariente** [comment]
Have you tried combining the SI-SDR loss with the spectral loss? Does the MSE on validation gets better?

Also, it's hard to debug without knowing the content of `stft` but with a normal STFT, I don't know why that wouldn't work.

**Reply by popcornell** [reply]
with lambda = 1 there is not any constraint over the phase, with lambda = 0 I don't know if decoupling the imag and real part makes sense

**Reply by jonashaag** [reply]
STFT is just `torch.stft`:

```py
@functools.lru_cache
def stft(dev, win=480):
    return functools.partial(
        torch.stft,
        n_fft=win,
        hop_length=win // 4,
        window=torch.hamming_window(win, device=dev),
        return_complex=True,
    )
```

> Have you tried combining the SI-SDR loss with the spectral loss?

When combining with SI-SDR it seems to work ok. (Not sure yet if results are better or worse than SI-SDR-only.)

> Does the MSE on validation gets better?

What kind of MSE do you mean? Time-domain MSE? The first picture is spectrogram validation loss (so, not a pure TF MSE but something similar).

**Reply by mpariente** [reply]
Re "Does the MSE on validation gets better?"

I meant to ask, if you train with SI-SDR, does your current loss also gets better? I'm trying to spot some instability issues, that could be solved with sample level losses. 

With lambda=1, it's classic mag masking, it should work without problem. 

Have you tried to compute the metrics with ideal masks?

**Comment by errorfixrepeat** [comment]
Does the model train smoothly when you use the same window size in SingleSrcMultiScaleSpectral() and alpha=0? (you'll need to use PIT training just to test, as it doesn't handle multichannel or produce a single valued output)

Torch's FFT expects (samples) or (batch, samples). Have you confirmed your input's are the right shape? If you have src along the second dimension you might get some very odd results.

**Reply by errorfixrepeat** [reply]
I found this in an old commit of open-unmix-pytorch  before they switched to asteroid filterbanks [src](https://www.programcreek.com/python/example/116151/torch.stft)
```
def forward(self, x):
        """
        Input: (nb_samples, nb_channels, nb_timesteps)
        Output:(nb_samples, nb_channels, nb_bins, nb_frames, 2)
        """

        nb_samples, nb_channels, nb_timesteps = x.size()

        # merge nb_samples and nb_channels for multichannel stft
        x = x.reshape(nb_samples*nb_channels, -1)

        # compute stft with parameters as close as possible scipy settings
        stft_f = torch.stft(
            x,
            n_fft=self.n_fft, hop_length=self.n_hop,
            window=self.window, center=self.center,
            normalized=False, onesided=True,
            pad_mode='reflect'
        )

        # reshape back to channel dimension
        stft_f = stft_f.contiguous().view(
            nb_samples, nb_channels, self.n_fft // 2 + 1, -1, 2
        )
        return stft_f 
```

**Comment by jonashaag** [comment, marked as answer]
Solved it! Take a look at those params over time (batch norm weights):

<img width="342" alt="Screenshot 2021-05-19 at 13 21 49" src="https://user-images.githubusercontent.com/175722/118804691-4dd74400-b8a5-11eb-8772-2fc96c896ce0.png">
<img width="335" alt="Screenshot 2021-05-19 at 13 21 45" src="https://user-images.githubusercontent.com/175722/118804692-4f087100-b8a5-11eb-909f-4085fad0af10.png">

The difference is that in the second training (good) I multiply the MSE by a large constant before computing the mean:

```py
c = 100_000  # good
c = 1  # bad
torch.mean((est_mag - tar_mag) ** 2 * c)
```

It also works ok if I multiply the estimates and targets by `c` before computing the MSE, either in the time domain before the STFT or in the TF domain.

The following works ok as well:

```py
# Good
c = 1
torch.sum((est_mag - tar_mag) ** 2 * c, dim=[1, 2]).mean()
```

But those do not work:

```py
# Bad
c = 1
torch.mean((est_mag - tar_mag + 1e-8) ** 2 * c)
torch.mean((est_mag - tar_mag) ** 2 * c + 1e-8)
torch.mean((est_mag - tar_mag) ** 2 * c, dim=[1, 2]).sum()
(torch.sum((est_mag - tar_mag) ** 2 * c, dim=[1, 2]) / est_mag[0].numel()).mean()
(torch.sum((est_mag - tar_mag) ** 2 * c / est_mag[0].numel(), dim=[1, 2])).mean()
```

**Reply by mpariente** [reply]
So, it seems to be a matter of machine precision. Have you tried to run the training with c=1, but in double?
